Use full path to update-alternatives on RedHat like systems

gfit-noc opened the pull request:
Patch to use the full path to `update-alternatives` to mitigate a "Command not found" error on a minimal CentOS 7 server.

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

jmlrt commented:
Hi @gfit-noc 
We couldn't reproduce your issue.

Can you give us more informations?

- What is your minimal CentOS 7 server source (docker image, ISO, AMI, ...)

- Can you run these 2 commands and paste  the results:

```
$ whoami
$ echo $PATH
```

gfit-noc commented:
We are running CentOS 7 minimal installed via Kickstart from ISO image. We do not manipulate the PATH variable.

I've added the following tasks before "Get the installed java path":

```
- shell: echo $PATH
  become: true
  register: pathvar

- debug:
    var: pathvar

- shell: whoami
  become: true
  register: whoamivar

- debug:
    var: whoamivar
```

Resulted in:

```
TASK [ansible-elasticsearch : shell] *******************************************************************************************************************************************************************************************************************************************
changed: [log1-dev.mgmt.t2.alldev]

TASK [ansible-elasticsearch : debug] *******************************************************************************************************************************************************************************************************************************************
ok: [log1-dev.mgmt.t2.alldev] => {
    "pathvar": {
        "changed": true,
        "cmd": "echo $PATH",
        "delta": "0:00:00.004860",
        "end": "2019-05-06 08:59:20.782807",
        "failed": false,
        "rc": 0,
        "start": "2019-05-06 08:59:20.777947",
        "stderr": "",
        "stderr_lines": [],
        "stdout": "/usr/local/bin:/usr/bin",
        "stdout_lines": [
            "/usr/local/bin:/usr/bin"
        ]
    }
}

TASK [ansible-elasticsearch : shell] *******************************************************************************************************************************************************************************************************************************************
changed: [log1-dev.mgmt.t2.alldev]

TASK [ansible-elasticsearch : debug] *******************************************************************************************************************************************************************************************************************************************
ok: [log1-dev.mgmt.t2.alldev] => {
    "whoamivar": {
        "changed": true,
        "cmd": "whoami",
        "delta": "0:00:00.009814",
        "end": "2019-05-06 08:59:21.503149",
        "failed": false,
        "rc": 0,
        "start": "2019-05-06 08:59:21.493335",
        "stderr": "",
        "stderr_lines": [],
        "stdout": "root",
        "stdout_lines": [
            "root"
        ]
    }
}
```

I guess using a full path to applications is never a bad idea.

jmlrt commented:
I think that there is a problem with your root's `PATH` variable.

After a fresh install of CentOS 7 using [CentOS-7-x86_64-Minimal-1810.iso](http://isoredirect.centos.org/centos/7/isos/x86_64/CentOS-7-x86_64-Minimal-1810.iso), root PATH should be `/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin` and not `/usr/local/bin:/usr/bin`.

![VirtualBox_test-centos_06_05_2019_10_48_54](https://user-images.githubusercontent.com/8582351/57215494-5249a080-6fed-11e9-9e13-dd1e7c8b5444.png)

Outside of that system tools like that are not in the same path in all distributions, for example `update-alternatives` is in `/usr/bin` directory in Debian and Ubuntu while it is in `/usr/sbin` directory in CentOS:

```$ docker run -it debian:9
root@8aab1ebe5a08:/# which update-alternatives
/usr/bin/update-alternatives
root@8aab1ebe5a08:/# exit

$ docker run -it ubuntu:18.04
root@9253c96adc19:/# which update-alternatives
/usr/bin/update-alternatives
root@9253c96adc19:/# exit

$ docker run -it centos:7
[root@55157738a640 /]# yum install -y -q which
warning: /var/cache/yum/x86_64/7/base/packages/which-2.20-7.el7.x86_64.rpm: Header V3 RSA/SHA256 Signature, key ID f4a80eb5: NOKEY
Public key for which-2.20-7.el7.x86_64.rpm is not installed
Importing GPG key 0xF4A80EB5:
 Userid     : "CentOS-7 Key (CentOS 7 Official Signing Key) <security@centos.org>"
 Fingerprint: 6341 ab27 53d7 8a78 a7c2 7bb1 24c6 a8a7 f4a8 0eb5
 Package    : centos-release-7-6.1810.2.el7.centos.x86_64 (@CentOS)
 From       : /etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7
install-info: No such file or directory for /usr/share/info/which.info.gz
[root@55157738a640 /]# which update-alternatives
/usr/sbin/update-alternatives```

For this reason, I prefer not using full path for in Ansible roles.

gfit-noc commented:
Alright. I close this pull request.
